File: 조준장/level2/택배 배달과 수거하기.py
def collection(pickups, cap):
    while pickups and cap > 0:
        pickup = pickups.pop()
        if pickup <= cap :
            cap -= pickup
        else:
            pickups.append(pickup - cap)
            break


def delivery(deliveries, cap):
    while deliveries and cap > 0:
        deliverie = deliveries.pop()
        if deliverie <= cap :
            cap -= deliverie
        else :
            deliveries.append(deliverie - cap)
            break
                
def solution(cap, n, deliveries, pickups):
    answer = 0
    while deliveries or pickups :
        while deliveries and deliveries[-1] == 0 :
            deliveries.pop()
        while pickups and pickups[-1] == 0:
            pickups.pop()
            
        answer += max(len(deliveries), len(pickups))*2
        
        delivery(deliveries, cap)
        collection(pickups, cap)
        
        
    return answer

File: 김찬우/프로그래머스 /Lv.2(미해결)/택배 배달과 수거하기.java
import java.util.ArrayList;
import java.util.*;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.lang.Math;


class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long result = 0;    
        
        List<List<Integer>> delivery= new ArrayList<>();
        List<List<Integer>> pickup = new ArrayList<>();
        
        for(int i = deliveries.length-1; i > -1 ;i--){
            
            if(deliveries[i]==0){
                continue;
                
            }else{
            
            List<Integer> list = new ArrayList<Integer>();
            list.add(i+1);
            list.add(deliveries[i]);
            delivery.add(list);
            }
         
        }
        
        for(int j = pickups.length-1; j > -1 ;j-- ){
            
            if(pickups[j]==0){
                continue;
            }
            
            List<Integer> list = new ArrayList<Integer>();
            list.add(j+1);
            list.add(pickups[j]);
            pickup.add(list);
            
        }
        
        while(!(delivery.size()==0 && pickup.size()==0)){
            
            long num = Math.max(delivery.get(delivery.size()-1),delivery.get(pickup.size()-1));
                                             
            bCap = cap;
            pCap = cap;
            
            
        }
        
      
        
        
        return result;
    }
}
